### PR TITLE
Revert "Revert "Cleaner ActiveRecordQueries mixin""

### PR DESCRIPTION
**Pull request by danwakefield**
Reverts gocardless/statesman#370

Readds these changes for a major release.

### CODE REVIEW

**Comment by ahjmorton** [comment]
Maybe make a pre-emptive uptick to the major version?

**Comment by danwakefield** [comment]
Going to leave them as separate PR's just to be safe.

File: lib/statesman/adapters/active_record_queries.rb [hunk 1]
```diff
@@ -1,51 +1,122 @@
...
+        ClassMethods.new(**args)
+      end
+
+      class ClassMethods < Module
+        def initialize(**args)
+          @args = args
+        end
+
+        def included(base)
+          ensure_inheritance(base)
```
Comment (davegudge) [inline lib/statesman/adapters/active_record_queries.rb:40]: Are there any docs to how the inheritance should be setup? 

I have the following setup:

```
    rails (~> 5.2.4)
    enumerize (2.3.1)
    statesman (7.2.0)
```


```ruby
class Candidate < ApplicationRecord
  extend Enumerize
  ...
  include Statesman::Adapters::ActiveRecordQueries[
    initial_state: :incomplete,
    transition_class: CandidateTransition
  ]
  has_many :candidate_transitions, inverse_of: :candidate, dependent: :destroy
  ...
  enumerize :state, in: CandidateStateMachine.states, default: :incomplete, predicates: true
  ...
end
```

```ruby
class Worker < Candidate
  ...
end
```

Upon instantiating a `Worker`, a stack error occurs:

```
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `block in ensure_inheritance'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/enumerize-2.3.1/lib/enumerize/base.rb:37:in `inherited'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `call'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `block in ensure_inheritance'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/enumerize-2.3.1/lib/enumerize/base.rb:37:in `inherited'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `call'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `block in ensure_inheritance'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/enumerize-2.3.1/lib/enumerize/base.rb:37:in `inherited'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `call'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `block in ensure_inheritance'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/enumerize-2.3.1/lib/enumerize/base.rb:37:in `inherited'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `call'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/statesman-7.2.0/lib/statesman/adapters/active_record_queries.rb:60:in `block in ensure_inheritance'
     # /home/travis/.rvm/gems/ruby-2.6.4/gems/enumerize-2.3.1/lib/enumerize/base.rb:37:in `inherited'
     # ./app/models/worker.rb:1:in `<top (required)>'
```

Removing the `ensure_inheritance(base)` line from ` lib/statesman/adapters/active_record_queries.rb` allows me to instantiate the `Worker` class as expected, and I can call the statesman methods on the model:

```
Worker.last.history
  Worker Load (0.5ms)  SELECT  "candidates".* FROM "candidates" ORDER BY "candidates"."id" DESC LIMIT $1  [["LIMIT", 1]]
  CandidateTransition Load (3.8ms)  SELECT "candidate_transitions".* FROM "candidate_transitions" WHERE "candidate_transitions"."candidate_id" = $1 ORDER BY "candidate_transitions"."sort_key" ASC  [["candidate_id", xxxx]]
=> [#<CandidateTransition:0x000000010dbaee20
  id: xxxx,
  candidate_id: xxxx,
  to_state: "active",
  metadata: {},
  sort_key: 10,
  created_at: Mon, 03 Aug 2020 11:05:28 UTC +00:00,
  created_by_id: xxxx,
  most_recent: true,
  updated_at: Mon, 03 Aug 2020 11:05:28 UTC +00:00>]
```

Comment (davegudge) [reply lib/statesman/adapters/active_record_queries.rb:40]: The [issue above](https://github.com/gocardless/statesman/pull/373#discussion_r465877454) is still occurring using:

```
    rails (7.0.4.3)
    enumerize (2.6.1)
    statesman (10.2.2)
```

I've opened https://github.com/gocardless/statesman/pull/511

**Pull request merged**